src/citrine/builders/design_spaces.py

```diff
@@ -235,3 +244,93 @@ def cartesian_join_design_spaces(
+                              ) -> DataSourceDesignSpace:
+    """[ALPHA] Convert an EnumeratedDesignSpace into a DataSourceDesignSpace.
+
+    Converts an EnumeratedDesignSpace into a DataSourceDesignSpace by writing
+    the data to a csv file, uploading it to the given dataset, and wrapping it
+    in a CSVDataSource.
+
+    Parameters
+    ----------
+    enumerated_ds: EnumeratedDesignSpace
+        data source to convert
+    dataset: Dataset
+        dataset into which to upload the data file
+    filename: Optional[str]
+        filename to use for the data file (default: design space name + "source data")
+
+    """
+
+    descriptors = {d.key: d for d in enumerated_ds.descriptors}
+    headers = [d.key for d in enumerated_ds.descriptors]
+
+    csv_filename = filename or "{} source data".format(enumerated_ds.name)
+    with open(csv_filename, "w", newline="") as csvfile:
+        writer = csv.writer(csvfile)
+        writer.writerow(headers)
+        for datum in enumerated_ds.data:
+            writer.writerow([datum.get(k, '') for k in headers])
+
+    file_link = dataset.files.upload(csv_filename)
+    data_source = CSVDataSource(file_link, descriptors)
+    data_source_ds = DataSourceDesignSpace(
+        name=enumerated_ds.name, description=enumerated_ds.description, data_source=data_source)
+    return data_source_ds
+
+
+def migrate_enumerated_design_space(*,
+                                    project: Project,
+                                    uid: UUID,
+                                    dataset: Union[str, Dataset],
+                                    filename: Optional[str],
+                                    cleanup: bool = True
+) -> DataSourceDesignSpace:
+    """
+    Migrate an EnumeratedDesignSpace on the Citrine Platform to a DataSourceDesignSpace
+    Parameters
+    ----------
+    project: Project
+        Project to use when accessing the Citrine Platform
+    uid: UUID
+        Unique identifier of the EnumeratedDesignSpace to migrate
+    dataset: Union[str, Dataset]
+        Dataset or name of dataset into which to write the data as a CSV.
+        If given as a string, will find_or_create using that name.
+    filename: Optional[str]
+        Optional string name to specify for the data CSV file.
+        Defaults to the design space name + "source data"
+    cleanup: bool
+        Whether or not to archive the migrated dataset if the migration is successful
+        Default: true
+
+    Returns
+    -------
+    DataSourceDesignSpace
+        The resulting dataset, which should have functional parity with the source design space.
+
+    """
+    enumerated_ds = project.design_spaces.get(uid)
+    if not isinstance(enumerated_ds, EnumeratedDesignSpace):
+        msg = "Trying to migrate an enumerated design space but this is a {}".format(type(enumerated_ds))
+        raise ValueError(msg)
+
+    if isinstance(dataset, str):
+        dataset = find_or_create_dataset(project.datasets, dataset)
```

Comment:
Why is this "create dataset" functionality included? Would we not expect the user to have a ready dataset in the common case?

Reply:
To make it even easier to use.  I agree that creating datasets just for this purpose is probably not the right behavior, though.  How about leaving the string option in and looking up the dataset by name if the string is given, but not creating the dataset if it doesn't already exist?

Reply:
These are utilities that are meant to benefit the user, so my inclination is to do the minimal thing that we know is useful. In this case, that would be to make the user specify the dataset itself, since the utility here is coming from the migration. If we find out that users want to supply a string and have it look up or create a dataset of that name, then we can add that in later.